### PR TITLE
Done

### PR DESCRIPTION
**Pull request by coltonstaab1**

### CODE REVIEW

**Comment by randallreedjr** [comment]
Hi @coltonstaab1. Unfortunately, there were some extra files included in your submission that make it impossible to review your code. Could you please remove the `.bundle` directory and resubmit?

**Comment by coltonstaab1** [comment]
Hi Randall,

Sorry about the delay--just updated my project, so it should be good to go.

Thanks,
Colton

**Comment by randallreedjr** [comment]
Hi @coltonstaab1, it looks like the project still has the .bundle directory.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -2,9 +2,126 @@
...
+	end
 
-end
\ No newline at end of file
+	get '/' do
+		erb :index
+	end
+
+	get '/signup' do
+		if !session[:id].nil?
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: I don't see a `current_user` or `logged_in?` helper method, but they can be helpful when dealing with authenticated users. A bit more semantic than looking up session variables.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,9 +2,126 @@
...
+			redirect '/tweets'
+		end
+		erb :'/users/create_user'
+	end
+
+	post '/signup' do
+		@user = User.create(params)
+		if params[:username] == "" || params[:password] == "" || params[:email] == ""
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: I think you could use [ActiveRecord validations](https://learn.co/tracks/full-stack-web-development/sinatra/activerecord/user-authentication-in-sinatra) (e.g. `validates_presence_of :username`) to perform these checks more easily.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -2,9 +2,126 @@
...
+			redirect '/signup'
+		end
+
+		if @user.save
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: `User.create` should have persisted the user for you. If you want to include an `if @user.save` check, `User.new` might make more sense.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -2,9 +2,126 @@
...
+		end
+	end
+
+	post "/login" do
+		@user = User.find_by(username: params[:username])
+		if @user && @user.authenticate(params[:password])
+			session[:id] = @user.id
+			redirect "/tweets"
+		else
+			redirect "/failure"
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: Where does this go?

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -2,9 +2,126 @@
...
+		end		
+		erb :'/tweets/new'
+	end
+
+	post '/tweets/new' do
+		if params[:content] == ""
+			redirect :'/tweets/new'	
+		end
+		@tweet = Tweet.create(params)
+		@tweet.user_id = session[:id]
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: The tweet creation and user association can both be accomplished in the same step:

``` ruby
@tweet = Tweet.new(content: params[:content], user: current_user)
```

or alternatively

```
@tweet = current_user.tweets.create(content: params[:content])
```

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -2,9 +2,126 @@
...
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		erb :'/tweets/display_single_tweet'
+	end
+
+	get '/tweets/:tweet_id/edit' do
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		if session[:id].nil?
+			redirect '/login'
+		end
+		if session[:id] != @tweet.user_id
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: I rarely see this check, well done!

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -2,9 +2,126 @@
...
+		if params[:content] != ""
+			@tweet.content = params[:content]
+			@tweet.save
+		end
+		#redirect '/tweets/:tweet_id'
+	end
+
+	post '/tweets/:tweet_id/delete' do
+		@tweet = Tweet.find_by_id(params[:tweet_id])
+		if session[:id] != @tweet.user_id
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:tweet_id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by`, like `find_by_id`, will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+class User < ActiveRecord::Base
+	has_secure_password
+	has_many :tweets
+
+	def slug
+		self.username.downcase.gsub(" ","-")
```
Comment (randallreedjr) [inline app/models/user.rb]: This is interesting...wouldn't it be possible to have multiple users with the same slug? For example '@john smith' and '@john-smith'.

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -2,9 +2,126 @@
...
+		@user = User.find_by(username: params[:username])
+		if @user && @user.authenticate(params[:password])
+			session[:id] = @user.id
+			redirect "/tweets"
+		else
+			redirect "/failure"
+		end
+	end
+
+	get '/tweets' do
```
Comment (randallreedjr) [inline app/controllers/application_controller.rb]: This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).

**Comment by randallreedjr** [comment]
@coltonstaab1 Sorry for the delay in reviewing this submission, glad we got the bundle files straightened out. You did a great job! I left a few notes, feel free to respond with any questions or comments. Closing this PR for now.

**Pull request closed without merging**